includes/model/class-openedx-woocommerce-plugin-enrollment.php

```diff
@@ -0,0 +1,785 @@
+                'private'                   => true,
+                'exclude_from_search'       => false,
+                'show_in_admin_all_list'    => true,
+                'show_in_admin_status_list' => true,
+                'label_count'               => _n_noop( 'Error <span class="count">(%s)</span>', 'Error <span class="count">(%s)</span>', 'wp-edunext-marketing-site' ),
+            )
+        );
+    }
+
+    /**
+     * Wrapper for the WP function that prevents an infinite cycle of hook recursion
+     *
+     * @param array $post The post info in an array.
+     */
+    public function wp_update_post( $post ) {
+        $this->unregister_save_hook();
+
+        wp_update_post( $post );
+
+        $this->register_save_hook();
+    }
+
+    /**
+     * Save post metadata when a post is saved.
+     *
+     * @param int  $post_id The post ID.
+     * @param post $post The post object.
+     * @param bool $update Whether this is an existing post being updated or not.
+     */
+    public function save_action( $post_id, $post, $update ) {
+
+        if ( $this->post_type !== $post->post_type ) {
+            return;
+        }
+
+        $oerarr = array(
+            'oer_course_id'    => sanitize_text_field( $_POST['oer_course_id'] ),
+            'oer_email'        => sanitize_text_field( $_POST['oer_email'] ),
+            'oer_username'     => sanitize_text_field( $_POST['oer_username'] ),
+            'oer_mode'         => sanitize_text_field( $_POST['oer_mode'] ),
+            'oer_request_type' => sanitize_text_field( $_POST['oer_request_type'] ),
+            'oer_order_id'     => sanitize_text_field( $_POST['oer_order_id'] ),
+        );
+
+        $oer_action = sanitize_text_field( $_POST['oer_action'] );
+
+        $this->save_eor( $post, $oerarr, $oer_action );
+    }
+
+    /**
+     * Creates a new post in the database and runs it through the save.
+     *
+     * @param array  $oerarr An array containing the eor info.
+     * @param string $oer_action The API action to perform once the wp process is done.
+     *
+     * @return object $post The post object.
+     */
+    public function insert_new( $oerarr, $oer_action = '' ) {
+        $this->unregister_save_hook();
+
+        $new_oer = array(
+            'post_content' => 'Created automatically by woocommerce to fullfill an order.',
+            'post_type'    => 'openedx_enrollment',
+        );
+        $post_id = wp_insert_post( $new_oer );
+        $post    = get_post( $post_id );
+
+        $this->save_eor( $post, $oerarr, $oer_action );
+        return $post;
+    }
+
+    /**
+     * Save openedx request based on the incomming args.
+     *
+     * @param post   $post The post object.
+     * @param array  $oerarr An array containing the eor info.
+     * @param string $oer_action The API action to perform once the wp process is done.
+     */
+    public function save_eor( $post, $oerarr, $oer_action ) {
+
```

Comment:
one thing we could refactor about this is that saving the post_type is dificult to test. Maybe we can do getter and setter methods that can be better unit tested. What is your opinion on this?

Reply:
Sounds good! Let me dig deeper into this for the next PR